ni_measurementlink_service/_drivers/__init__.py

```diff
@@ -4,25 +4,52 @@
 import contextlib
 from typing import ContextManager, TypeVar
 
+from ni_measurementlink_service.session_management._types import SessionInitializationBehavior
+
 TSession = TypeVar("TSession")
 
+_EXISTING_BEHAVIORS = [
+    SessionInitializationBehavior.AUTO,
+    SessionInitializationBehavior.ATTACH_TO_SERVER_SESSION,
+    SessionInitializationBehavior.INITIALIZE_SERVER_SESSION,
+]
+
 
-def closing_session(session: TSession) -> ContextManager[TSession]:
+def closing_session(
+    session: TSession,
+    initialization_behavior: SessionInitializationBehavior = SessionInitializationBehavior.AUTO,
+    can_emulate: bool = True,
+) -> ContextManager[TSession]:
     """Create a context manager that closes the session.
 
     Args:
         session: A driver session.
 
+        initialization_behavior: Specifies whether to initialize a new
+            session or attach to an existing session.
+
+        can_emulate: Specify True, if MeasurementLink must emulate the
+            INITIALIZE_SERVER_SESSION_THEN_DETACH and
+            ATTACH_TO_SERVER_SESSION_THEN_CLOSE behavior. Else, False.
+
     Returns:
         A context manager that yields the session and closes it.
+
+    Raises:
+        TypeError: If the session is not a context manager and if it does not have a close() method.
+
+        ValueError: If the initialization behavior is invalid.
     """
-    if isinstance(session, contextlib.AbstractContextManager):
-        # Assume the session yields itself.
+    if not hasattr(session, "close"):
+        raise TypeError("Session must have a close() method.")
+    elif initialization_behavior not in SessionInitializationBehavior:
+        raise ValueError(f"Invalid initialization behavior: '{initialization_behavior}'.")
+
+    if not can_emulate or initialization_behavior in _EXISTING_BEHAVIORS:
+        if not isinstance(session, contextlib.AbstractContextManager):
+            raise TypeError("Session must be a context manager.")
         return session
-    elif hasattr(session, "close"):
-        return contextlib.closing(session)
-    else:
-        raise TypeError(
-            f"Invalid session type '{type(session)}'. A session must be a context manager and/or "
-            "have a close() method."
-        )
+
+    if initialization_behavior == SessionInitializationBehavior.INITIALIZE_SESSION_NO_CLOSE:
+        return contextlib.nullcontext(session)
+    return contextlib.closing(session)
```

Comment:
use elif

Reply:
If I add an `elif` case, then "mypy" insists on adding an else case but logically all the enum values are covered until or within the elif case itself (if I add one), so adding an else case seemed redundant. That was the reason to return "contextlib.closing(session)" directly. 

So, would you still recommend having an `elif` case and then within the `else` block, return the "session" object or another `return contextlib.closing(session)` statement?

Reply:
If you validate initialization_behavior based on the cases covered by the function, it should work out. This is better because it does the right thing if we add an enum value without updating this function.

```python
    if initialization_behavior in _EXISTING_BEHAVIORS:
        if not isinstance(session, contextlib.AbstractContextManager):
            raise TypeError("Session must be a context manager.")
        return session
    elif initialization_behavior == SessionInitializationBehavior.INITIALIZE_SESSION_THEN_DETACH:
        return contextlib.nullcontext(session)
    elif initialization_behavior == SessionInitializationBehavior.ATTACH_SESSION_THEN_CLOSE:
        return contextlib.closing(session)
    else:
        raise ValueError(f"Invalid initialization behavior: '{initialization_behavior}'.")
```